cmd/snap-confine/snap-confine-invocation.c

```diff
@@ -0,0 +1,88 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+#include "snap-confine-invocation.h"
+
+#include <stdlib.h>
+#include <string.h>
+
+#include "../libsnap-confine-private/cleanup-funcs.h"
+#include "../libsnap-confine-private/snap.h"
+#include "../libsnap-confine-private/string-utils.h"
+#include "../libsnap-confine-private/utils.h"
+
+void sc_init_invocation(sc_invocation *inv, const struct sc_args *args, const char *snap_instance) {
+    /* Snap instance name is conveyed via untrusted environment. It may be
+     * unset (typically when experimenting with snap-confine by hand). It
+     * must also be a valid snap instance name. */
+    if (snap_instance == NULL) {
+        die("cannot use NULL snap instance name");
+    }
+    sc_instance_name_validate(snap_instance, NULL);
+
+    /* The security tag is conveyed via untrusted command line. It must be
+     * in agreement with snap instance name and must be a valid security
+     * tag. */
+    const char *security_tag = sc_args_security_tag(args);
+    if (!verify_security_tag(security_tag, snap_instance)) {
+        die("security tag %s not allowed", security_tag);
+    }
+
+    /* The base snap name is conveyed via untrusted, optional, command line
+     * argument. It may be omitted where it implies the "core" snap is the
+     * base. */
+    const char *base_snap_name = sc_args_base_snap(args);
+    if (base_snap_name == NULL) {
+        base_snap_name = "core";
+    }
+    sc_snap_name_validate(base_snap_name, NULL);
+
+    /* The executable is conveyed via untrusted command line. It must be set
+     * but cannot be validated further than that at this time. It might be
+     * arguable to validate it to be snap-exec in one of the well-known
+     * locations or one of the special-cases like strace / gdb but this is
+     * not done at this time. */
+    const char *executable = sc_args_executable(args);
+    if (executable == NULL) {
+        die("cannot run with NULL executable");
+    }
+
+    /* Invocation helps to pass relevant data to various parts of snap-confine. */
+    memset(inv, 0, sizeof *inv);
+    inv->base_snap_name = sc_strdup(base_snap_name);
+    inv->executable = sc_strdup(executable);
+    inv->security_tag = sc_strdup(security_tag);
+    inv->snap_instance = sc_strdup(snap_instance);
+    inv->classic_confinement = sc_args_is_classic_confinement(args);
+
+    debug("security tag: %s", inv->security_tag);
+    debug("executable:   %s", inv->executable);
+    debug("confinement:  %s", inv->classic_confinement ? "classic" : "non-classic");
+    debug("base snap:    %s", inv->base_snap_name);
+}
+
+static void sc_fini_invocation(sc_invocation *inv) {
+    sc_cleanup_string(&inv->snap_instance);
+    sc_cleanup_string(&inv->base_snap_name);
+    sc_cleanup_string(&inv->security_tag);
+    sc_cleanup_string(&inv->executable);
+}
+
+void sc_cleanup_invocation(sc_invocation *inv) {
+    if (inv != NULL) {
+        sc_fini_invocation(inv);
```

Comment:
I don't see why we do not just fold sc_fini_invocation() here? I mean, this helper is only used in a single place, its private and trivial. Whats the advantage of the split? Will sc_fini_invocation be used in more places in the future? Or is it just a code-style preference?

Reply:
It's a code style preference. I know it's a hit-and-miss so far (mainly because it is a while to change something in that codebase). I would like to have init/fini for the contents of the structure but not for the structure itself.

Given that we are getting into a flow of refactoring snap-confine I would like to keep it, so that it can be in line with other structures that get the same treatment.

Reply:
What is the difference between `sc_cleanup_...` and `sc_fini_...` ? I still don't see why we need two when they do the same? Maybe I need an example where the two are actually different? If we want symetry, why not "init" and "cleanup"? Sorry for being so thick here but I have this YAGNI feeling and I really would like to understand if its real or just me not knowing the details.

Reply:
Cleanup is specifically the cleanup handler that takes a pointer to the object being cleaned up (e.g. a `int *` for sc_cleanup_close`). The finaliser runs on the object with releasing the memory of the object itself (e.g. `close` as an analogy). In this case however those are identical because the structure is not heap allocated.

Reply:
After discussing this I folded the function into one and decided to make a separate pass to make memory management more consistent around the tree.